In [ ]:
# DS4400 HW 2
# Eunchae Hong
# Problem 2: Linear Regression
import csv
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import *

house_data = pd.read_csv("kc_house_data.csv")
train_data = pd.read_csv("train.csv")
test_data = pd.read_csv("test.csv")


In [ ]:
# Problem 2 Part 1
# Report the coefficients of the linear regression models and the following metrics on the training data: (1) MSE metric; (2) R^2 Metric

# train the multiple linear regrssion model
house_features = house_data.drop(columns = ["id", "date", "zipcode", "price"])
house_price = house_data["price"]

X_train = train_data.drop(columns = ["zipcode", "price"])
y_train = train_data["price"]

X_test = test_data.drop(columns = ["id", "date", "zipcode", "price"])
y_test = test_data["price"]

# scale the data using StandardScalar
scale = StandardScaler()
X_train_scaled = scale.fit_transform(X_train)

model_train = LinearRegression()
model_train.fit(X_train_scaled, y_train)

y_train_pred = model_train.predict(X_train_scaled)

# report the coefficients of the linear regression models, the highest coefficient, MSE metric, and R^2 metric
coef = model_train.coef_
print("The coefficients of the linear regression models are:", coef)

feature_names = X_train.columns
max_idx = np.argmax(np.abs(coef)), 2

print("Highest coefficient feature:", feature_names[max_idx])
print("Coefficient value:", coef[max_idx])

train_mse = round(mean_squared_error(y_train, y_train_pred), 2)
print("The MSE metric for the training set is:", train_mse)

train_r_squared = round(r2_score(y_train, y_train_pred), 2)
print("The R Squared Metric for the training set it is:", train_r_squared)

The coefficients of the linear regression models are: [  8456.02390293 -12807.33864205  18456.91324953  57161.58157168
  11127.3380784    8151.03831585  64230.91137517  47610.28800795
  12647.60909652  92511.0761025   48439.05104545  27688.81217988
 -68043.17344882  17341.92574432  78129.85226103  -1437.66937441
  45479.12775574 -12906.56027599]
Highest coefficient feature: grade
Coefficient value: 92511.07610250056
The MSE metric for the training set is: 31415747916.1
The R Squared Metric for the training set it is: 0.73


In [78]:
# Problem 2 Part 2
# Evaluate the model on the testing set. Report the MSE and R^2 metrics on the testing set.

# scale = StandardScaler()
X_test_scaled = scale.transform(X_test)

y_test_pred = model_train.predict(X_test_scaled)

mse = round(mean_squared_error(y_test, y_test_pred), 2)
print("The MSE metric for the testing set is:", mse)

r_squared = round(r2_score(y_test, y_test_pred), 2)
print("The R Squared Metric for the testing set is:", r_squared)

The MSE metric for the testing set is: 58834673978.21
The R Squared Metric for the testing set is: 0.65


### Problem 2 Part 3 ###
Interpret the results in your own words. Which features contribute mostly to the linear regression model? Is the model fitting the data well? How large is the model error? How do the training and testing MSE relate?

The feature that contributes the most to the linear regression model is the one with the highest absolute value for its coefficient after standardizing the values. In this case the feature with the highest coefficent was grade with a coefficient of 92511. The grade feature is an index feature that determines the construction and design of the home. It is understandable that this feature would have the largest effect on the model as intricate construction and design of the home can impact how consumers view a home, therefore increasing price. 

The model demonstrated similar results with the training and testing set, therefore we can conclude the model is fitting the data relativley well and there is nothing striking as underfitting or overfitting within the data. 

The traning data has a MSE of 31,415,747,916.10, meaning that the average squared differences between the actual housing prices and the predicted hosing prices is around $31 billion. This is expected as housing prices are large and errors would be larger in this case as they are squared. Additionally an $R^2$ of 0.73, means that 73% of the variability in the housing prices can be explained by the model. 

The testing data has a MSE of 58,834,673,978.21 meaning that the average squared differences between the actual housing prices and the predicted housing prices around $58 billion. Additionally an $R^2$ of 0.65 means that 65% of the variability in the housing prices can be explained by the model. 

There is a large MSE becuase the housing prices generally large in itself, since the underlying data has large values and its errors are squared, this can explain why the MSE is so high here. Squaring also penalizing the larger errors more heavily, which is an aspect of MSE. 

Training and testing MSE relate becuase they are both measuring the mean squred differences between actual housing prices and predicted housing prcies, just on two different sets of data - training vs testing.